File: src/python/multiprocessing/05_queue_communication.py
"""
This script demonstrates the use of the multiprocessing module in Python to establish communication
between multiple processes using a Queue. The Queue allows for safe inter-process communication
by providing a simple way to pass data between processes. In this example, we create a producer process
that generates data and puts it into the Queue, and a consumer process that reads the data from the Queue.

When working with multiple processes, it is important to ensure that data is safely shared between them.
The multiprocessing module provides a Queue class that can be used to achieve this. The Queue uses a
lock to ensure that only one process can access it at a time, preventing race conditions and data corruption.

In this example, we demonstrate the use of a Queue to share data between a producer and a consumer process.
The producer generates data and puts it into the Queue, while the consumer reads the data from the Queue and
processes it. This pattern is useful when the producer and consumer processes work at different rates or when
you want to separate data generation from data processing.
"""

import multiprocessing
import time
import random

def producer(queue):
    for i in range(5):
        item = random.randint(1, 10)
        print(f"Producer: Adding {item} to the queue")
        queue.put(item)
        time.sleep(random.uniform(0.5, 1))

def consumer(queue):
    while True:
        item = queue.get()
        if item is None:
            break

        print(f"Consumer: Processing {item}")
        time.sleep(random.uniform(0.5, 1))

def main():
    # Create a Queue to share data between the producer and consumer processes
    queue = multiprocessing.Queue()

    # Create the producer and consumer processes
    producer_process = multiprocessing.Process(target=producer, args=(queue,))
    consumer_process = multiprocessing.Process(target=consumer, args=(queue,))

    # Start the producer and consumer processes
    producer_process.start()
    consumer_process.start()

    # Wait for the producer process to finish
    producer_process.join()

    # Add a sentinel value to the queue to signal the consumer process to exit
    queue.put(None)

    # Wait for the consumer process to finish
    consumer_process.join()

if __name__ == "__main__":
    main()

File: src/cpp/multiprocessing/05_queue_communication.cpp
/*
This program demonstrates the use of POSIX message queues to establish
communication between two processes in C++. A message queue allows for safe
inter-process communication by providing a simple way to pass data between
processes. In this example, we create a producer process that generates data and
writes it into the message queue, and a consumer process that reads the data
from the message queue.

When working with multiple processes, it is important to ensure that data is
safely shared between them. Message queues provide a way to achieve this. They
can be used to send data between a producer and a consumer process. The producer
generates data and writes it into the message queue, while the consumer reads
the data from the message queue and processes it. This pattern is useful when
the producer and consumer processes work at different rates or when you want to
separate data generation from data processing.
*/

#include <cstdlib>
#include <ctime>
#include <fcntl.h>
#include <iostream>
#include <mqueue.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>

const char *MQ_NAME = "/my_message_queue";

void producer() {
  srand(time(NULL));

  mqd_t mq = mq_open(MQ_NAME, O_WRONLY | O_CREAT, 0666, nullptr);
  if (mq == (mqd_t)-1) {
    perror("mq_open");
    exit(EXIT_FAILURE);
  }

  for (int i = 0; i < 5; ++i) {
    int item = rand() % 10 + 1;
    std::cout << "Producer: Adding " << item << " to the message queue"
              << std::endl;
    mq_send(mq, reinterpret_cast<char *>(&item), sizeof(item), 0);
    usleep((rand() % 500000) + 500000);
  }

  mq_close(mq);
}

void consumer() {
  mqd_t mq = mq_open(MQ_NAME, O_RDONLY);
  if (mq == (mqd_t)-1) {
    perror("mq_open");
    exit(EXIT_FAILURE);
  }

  int item;
  ssize_t bytesRead;
  while ((bytesRead = mq_receive(mq, reinterpret_cast<char *>(&item),
                                 sizeof(item), nullptr)) > 0) {
    std::cout << "Consumer: Processing " << item << std::endl;
    usleep((rand() % 500000) + 500000);
  }

  mq_close(mq);
}

int main() {
  pid_t pid = fork();
  if (pid == -1) {
    perror("fork");
    exit(EXIT_FAILURE);
  }

  if (pid == 0) {
    // Child process (Consumer)
    consumer();
    exit(EXIT_SUCCESS);
  } else {
    // Parent process (Producer)
    producer();

    // Wait for the child process to finish
    wait(NULL);
  }

  // Cleanup
  mq_unlink(MQ_NAME);

  return 0;
}

// g++ -o main main.cpp -lrt
